helper/convert/value.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package convert
+
+import (
+	"github.com/hashicorp/go-cty/cty"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+func PrimitiveTfValue(in cty.Value) tftypes.Value {
+	if in.IsNull() {
+		return emptyTfValue(ToTfType(in.Type()))
+	}
+
+	var val tftypes.Value
+	switch in.Type() {
+	case cty.String:
+		val = tftypes.NewValue(tftypes.String, in.AsString())
+	case cty.Bool:
+		val = tftypes.NewValue(tftypes.Bool, in.True())
+	case cty.Number:
+		val = tftypes.NewValue(tftypes.Number, in.AsBigFloat())
+	}
+
+	return val
+}
+
+func ListTfValue(in cty.Value) tftypes.Value {
+	listType := ToTfType(in.Type())
+
+	if in.IsNull() || in.LengthInt() == 0 {
+		return emptyTfValue(listType)
+	}
+
+	vals := make([]tftypes.Value, 0)
+
+	for _, v := range in.AsValueSlice() {
+		vals = append(vals, ToTfValue(v))
+	}
+
+	return tftypes.NewValue(listType, vals)
+}
+
+func MapTfValue(in cty.Value) tftypes.Value {
+	mapType := ToTfType(in.Type())
+
+	if in.IsNull() || in.LengthInt() == 0 {
+		return emptyTfValue(mapType)
+	}
+
+	vals := make(map[string]tftypes.Value)
+
+	for k, v := range in.AsValueMap() {
+		vals[k] = ToTfValue(v)
+	}
+
+	return tftypes.NewValue(mapType, vals)
+}
+
+func SetTfValue(in cty.Value) tftypes.Value {
+	setType := ToTfType(in.Type())
+
+	if in.IsNull() || in.LengthInt() == 0 {
+		return emptyTfValue(setType)
+	}
+
+	vals := make([]tftypes.Value, 0)
+
+	for _, v := range in.AsValueSlice() {
+		vals = append(vals, ToTfValue(v))
+	}
+
+	return tftypes.NewValue(setType, vals)
+}
+
+func ObjectTfValue(in cty.Value) tftypes.Value {
+	objType := ToTfType(in.Type())
+
+	if in.IsNull() || in.LengthInt() == 0 {
+		return emptyTfValue(objType)
+	}
+
+	vals := make(map[string]tftypes.Value)
+
+	for k, v := range in.AsValueMap() {
+		vals[k] = ToTfValue(v)
+	}
+
+	return tftypes.NewValue(objType, vals)
+}
+
+func TupleTfValue(in cty.Value) tftypes.Value {
+	tupleType := ToTfType(in.Type())
+
+	if in.IsNull() || in.LengthInt() == 0 {
+		return emptyTfValue(tupleType)
+	}
+
+	vals := make([]tftypes.Value, 0)
+
+	for _, v := range in.AsValueSlice() {
+		vals = append(vals, ToTfValue(v))
+	}
+
+	return tftypes.NewValue(tupleType, vals)
+}
+
+func ToTfValue(in cty.Value) tftypes.Value {
```

Comment:
Similar question to `ToTfType`, do these functions needs to be exported since they are all accessed via `ResourceData`? If not, it might make sense to just move this to the internal `plugin/convert` package.

If they do need to be exported, we should add some package docs 🙂

Reply:
It was intentional to add these as public functions since this type of `cty` -> `tftypes` conversion doesn't really exist anywhere else (at least not publicly) but since we want to shift away from extending SDKv2 functionality I felt like this could be a useful utility/puzzle piece for provider developers to have access to if they're wanting to mix and match things from SDKv2 and Framework (and even plugin-go?) even though the use cases might be quite rare. 

There's obviously quite a lot of speculation in that view though, perhaps this is something that we should hide away and only expose in a controlled manner. I'm not precious about it in any case.

WDYT?